src/replicache.ts

```diff
@@ -818,13 +865,34 @@ export class Replicache<MD extends MutatorDefs = {}>
     }
   }
 
-  private async _fireOnChange(): Promise<void> {
+  private async _fireOnChange(changedKeys: ChangedKeysMap): Promise<void> {
+    const subscriptions = subscriptionsForChangedKeys(
+      this._subscriptions,
+      changedKeys,
+    );
+    await this._fireSubscriptions(subscriptions);
+  }
+
+  private async _indexDefinitionChanged(name: string): Promise<void> {
+    // When an index definition changes we fire all subscriptions that uses
+    // index scans with that index.
+    const subscriptions = subscriptionsForIndexDefinitionChanged(
+      this._subscriptions,
+      name,
+    );
+    await this._fireSubscriptions(subscriptions);
+    this._broadcastChange(await this._root, new Map(), name);
+  }
+
+  private async _fireSubscriptions(
+    subscriptions: Iterable<Subscription<JSONValue | undefined, unknown>>,
+  ) {
+    const subs = [...subscriptions];
     type R =
       | {ok: true; value: JSONValue | undefined}
       | {ok: false; error: unknown};
-    const subscriptions = [...this._subscriptions];
     const results = await this.query(async tx => {
-      const promises = subscriptions.map(async s => {
+      const promises = subs.map(async s => {
         // Tag the result so we can deal with success vs error below.
         try {
           return {ok: true, value: await s.body(tx)} as R;
```

Comment:
This is not right. We need to update the watched keys here. I will add tests and implement